webpush/encrypt.go

```diff
@@ -250,15 +262,14 @@ func newInfo(infoType string, context []byte) []byte {
 //
 // This is a partial implementation of HKDF tailored to our specific purposes.
 // In particular, for us the value of N will always be 1, and thus T always
-// equals HMAC-Hash(PRK, info | 0x01).
+// equals HMAC-Hash(PRK, info | 0x01). This is true because the maximum output
+// length we need/allow is 32.
 //
 // See https://www.rfc-editor.org/rfc/rfc5869.txt
-func hkdf(salt, ikm, info []byte, length int) []byte {
+func hkdf(salt, ikm, info []byte, length int) ([]byte, error) {
 	// HMAC length for SHA256 is 32 bytes, so that is the maximum result length.
-	// HKDF defines a way to have longer lengths, but our partial implementation
-	// will just truncate the result.
 	if length > 32 {
-		length = 32
+		return nil, fmt.Errorf("Can only produce HKDF outputs up to 32 bytes long")
```

Comment:
So, "real" HKDF supports much longer output lengths, but since web push never needs those lengths we can simplify to the version we have here.

If you request a length of, say, 48 bytes and we don't reject that then the current function tries to return the first 48 bytes of a 32 byte array.

As you point out, we control all the callers and they definitely don't request longer outputs. Would you prefer that we just let this function cause a panic? I'm not really bothered either way, I just think that the current truncation could lead to us accidentally introducing a bug in the future. If the spec changes to require 48 byte keys then truncating might not fail obviously, whereas an error will be spotted.